# Lab 04: Joining

## #1
You should have already downloaded the [planes data](https://raw.githubusercontent.com/uc-python/intro-python-datasci/master/data/planes.csv).
Now download a related dataset – [flights data](https://raw.githubusercontent.com/uc-python/intro-python-datasci/master/data/flights.csv).

Note that this is a fairly large dataset and may take a few seconds to download.

## #2
Load in the planes data as `planes` and the flights data as `flights`.

In [1]:
import pandas as pd

In [3]:
planes = pd.read_csv('data/planes.csv')

In [4]:
flights = pd.read_csv('data/flights.csv')

In [5]:
planes.head()

tailnum    year                     type      manufacturer      model  \
0  N10156  2004.0  Fixed wing multi engine           EMBRAER  EMB-145XR   
1  N102UW  1998.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
2  N103US  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
3  N104UW  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
4  N10575  2002.0  Fixed wing multi engine           EMBRAER  EMB-145LR   

   engines  seats  speed     engine  
0        2     55    NaN  Turbo-fan  
1        2    182    NaN  Turbo-fan  
2        2    182    NaN  Turbo-fan  
3        2    182    NaN  Turbo-fan  
4        2     55    NaN  Turbo-fan

In [6]:
flights.head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute            time_hour  
0      1400     5      15  2013-01-01 05:00:00  
1      1416     5      29  2013-01-01 05:00:00  
2      1089     5      40  2013-01-01 05:00:00  
3      1576     5      45  2013-01-01 05:00:00  
4       762     6       0  2013-01-01 06:00:00

## #3
Based on what you see in the flights dataset, form a hypothesis about what this dataset is.
Do you see a column in this data that could match up with a column in the planes data?

The flights dataset provides details of numerous flight schedules which contains information about their origin, destination, duration, distance, date, delay, etc. The common key with planes dataset in the variable 'tailnum'

## #4
Just using the flights data, is it possible to determine what percentage of the flights were completed on Boeing planes?
Why not?

Only the flights dataset will be insufficient to provide information about BOEING manufacturered flights since it is not available. We must join it with planes data to get the required column.

## #5
In order to combine this data with the planes data (where the manufacturer column is), we need to do a join.

Inner join the planes data to the flights data, storing the result in a variable called `joined_data`.

In [7]:
joined_data = pd.merge(flights, planes, on = 'tailnum', how = 'inner')

## #6
How many rows are in the result DataFrame?
How does this compare to how many rows were in its parent DataFrames (planes and flights)?

In [9]:
joined_data.shape[0]

284170

The joined dataset contains fewer rows than flights since it contains data for only those tailnum which are present in planes dataset.

## #7
What would happen if you did a left join instead of an inner join?
Try it, storing the result in a variable called `left_joined_data`.
Check how many rows are in this new DataFrame -- is it different to the number of rows in `joined_data` (from #5)?

In [10]:
left_joined_data = pd.merge(flights, planes, on = 'tailnum', how = 'left')

In [12]:
left_joined_data.shape[0]

336776

Left join gives the same number of rows as in flights. This is because left join includes all data from the left table.

## #8
Using the original joined data, calculate what percentage of flights were completed on Boeing planes.

In [13]:
round(joined_data.loc[joined_data['manufacturer'] == 'BOEING'].shape[0]/joined_data.shape[0], 4)

0.2918

There are 29.18% of Boeing flights in the joined dataset.